src/monster.cpp

```diff
@@ -645,9 +645,16 @@ bool Monster::selectTarget(Creature* creature)
 		return false;
 	}
 
-	if (isHostile() || isSummon()) {
-		if (setAttackedCreature(creature) && !isSummon()) {
+	if (isSummon()) {
+		if (!canAttack(creature)) {
+			removeAttackedCreature();
+		}
+	} else if (isHostile()) {
+		if (canAttack(creature)) {
+			setAttackedCreature(creature);
 			g_dispatcher.addTask([id = getID()]() { g_game.checkCreatureAttack(id); });
+		} else {
+			removeAttackedCreature();
```

Comment:
This part was complex to understand. Internally, `setAttackedCreature` would remove the target and return a false when the current creature couldn’t attack it (e.g., due to `z != z` and visibility checks with canSee).